// coc-webview/lib/client.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};
var __commonJS = (cb, mod) => function __require2() {
  return mod || (0, cb[Object.keys(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};

// node_modules/socket.io-client/dist/socket.io.min.js
var require_socket_io_min = __commonJS({
  "node_modules/socket.io-client/dist/socket.io.min.js"(exports2, module2) {
    !function(t, e) {
      typeof exports2 == "object" && typeof module2 == "object" ? module2.exports = e() : typeof define == "function" && define.amd ? define([], e) : typeof exports2 == "object" ? exports2.io = e() : t.io = e();
    }(self, function() {
      return function(t) {
        var e = {};
        function n(r) {
          if (e[r])
            return e[r].exports;
          var o = e[r] = { i: r, l: false, exports: {} };
          return t[r].call(o.exports, o, o.exports, n), o.l = true, o.exports;
        }
        return n.m = t, n.c = e, n.d = function(t2, e2, r) {
          n.o(t2, e2) || Object.defineProperty(t2, e2, { enumerable: true, get: r });
        }, n.r = function(t2) {
          typeof Symbol != "undefined" && Symbol.toStringTag && Object.defineProperty(t2, Symbol.toStringTag, { value: "Module" }), Object.defineProperty(t2, "__esModule", { value: true });
        }, n.t = function(t2, e2) {
          if (1 & e2 && (t2 = n(t2)), 8 & e2)
            return t2;
          if (4 & e2 && typeof t2 == "object" && t2 && t2.__esModule)
            return t2;
          var r = Object.create(null);
          if (n.r(r), Object.defineProperty(r, "default", { enumerable: true, value: t2 }), 2 & e2 && typeof t2 != "string")
            for (var o in t2)
              n.d(r, o, function(e3) {
                return t2[e3];
              }.bind(null, o));
          return r;
        }, n.n = function(t2) {
          var e2 = t2 && t2.__esModule ? function() {
            return t2.default;
          } : function() {
            return t2;
          };
          return n.d(e2, "a", e2), e2;
        }, n.o = function(t2, e2) {
          return Object.prototype.hasOwnProperty.call(t2, e2);
        }, n.p = "", n(n.s = 18);
      }([function(t, e) {
        t.exports = typeof self != "undefined" ? self : typeof window != "undefined" ? window : Function("return this")();
      }, function(t, e, n) {
        var r = n(24), o = n(25), i = String.fromCharCode(30);
        t.exports = { protocol: 4, encodePacket: r, encodePayload: function(t2, e2) {
          var n2 = t2.length, o2 = new Array(n2), s = 0;
          t2.forEach(function(t3, c) {
            r(t3, false, function(t4) {
              o2[c] = t4, ++s === n2 && e2(o2.join(i));
            });
          });
        }, decodePacket: o, decodePayload: function(t2, e2) {
          for (var n2 = t2.split(i), r2 = [], s = 0; s < n2.length; s++) {
            var c = o(n2[s], e2);
            if (r2.push(c), c.type === "error")
              break;
          }
          return r2;
        } };
      }, function(t, e, n) {
        function r(t2) {
          if (t2)
            return function(t3) {
              for (var e2 in r.prototype)
                t3[e2] = r.prototype[e2];
              return t3;
            }(t2);
        }
        t.exports = r, r.prototype.on = r.prototype.addEventListener = function(t2, e2) {
          return this._callbacks = this._callbacks || {}, (this._callbacks["$" + t2] = this._callbacks["$" + t2] || []).push(e2), this;
        }, r.prototype.once = function(t2, e2) {
          function n2() {
            this.off(t2, n2), e2.apply(this, arguments);
          }
          return n2.fn = e2, this.on(t2, n2), this;
        }, r.prototype.off = r.prototype.removeListener = r.prototype.removeAllListeners = r.prototype.removeEventListener = function(t2, e2) {
          if (this._callbacks = this._callbacks || {}, arguments.length == 0)
            return this._callbacks = {}, this;
          var n2, r2 = this._callbacks["$" + t2];
          if (!r2)
            return this;
          if (arguments.length == 1)
            return delete this._callbacks["$" + t2], this;
          for (var o = 0; o < r2.length; o++)
            if ((n2 = r2[o]) === e2 || n2.fn === e2) {
              r2.splice(o, 1);
              break;
            }
          return r2.length === 0 && delete this._callbacks["$" + t2], this;
        }, r.prototype.emit = function(t2) {
          this._callbacks = this._callbacks || {};
          for (var e2 = new Array(arguments.length - 1), n2 = this._callbacks["$" + t2], r2 = 1; r2 < arguments.length; r2++)
            e2[r2 - 1] = arguments[r2];
          if (n2) {
            r2 = 0;
            for (var o = (n2 = n2.slice(0)).length; r2 < o; ++r2)
              n2[r2].apply(this, e2);
          }
          return this;
        }, r.prototype.listeners = function(t2) {
          return this._callbacks = this._callbacks || {}, this._callbacks["$" + t2] || [];
        }, r.prototype.hasListeners = function(t2) {
          return !!this.listeners(t2).length;
        };
      }, function(t, e, n) {
        var r = n(0);
        t.exports.pick = function(t2) {
          for (var e2 = arguments.length, n2 = new Array(e2 > 1 ? e2 - 1 : 0), r2 = 1; r2 < e2; r2++)
            n2[r2 - 1] = arguments[r2];
          return n2.reduce(function(e3, n3) {
            return t2.hasOwnProperty(n3) && (e3[n3] = t2[n3]), e3;
          }, {});
        };
        var o = setTimeout, i = clearTimeout;
        t.exports.installTimerFunctions = function(t2, e2) {
          e2.useNativeTimers ? (t2.setTimeoutFn = o.bind(r), t2.clearTimeoutFn = i.bind(r)) : (t2.setTimeoutFn = setTimeout.bind(r), t2.clearTimeoutFn = clearTimeout.bind(r));
        };
      }, function(t, e, n) {
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function i(t2, e2) {
          return (i = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function s(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = u(t2);
            if (e2) {
              var o2 = u(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return c(this, n2);
          };
        }
        function c(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return a(t2);
        }
        function a(t2) {
          if (t2 === void 0)
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
          return t2;
        }
        function u(t2) {
          return (u = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var f = n(1), l = n(2), p = n(3).installTimerFunctions, h = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && i(t3, e3);
          }(u2, t2);
          var e2, n2, r2, c2 = s(u2);
          function u2(t3) {
            var e3;
            return function(t4, e4) {
              if (!(t4 instanceof e4))
                throw new TypeError("Cannot call a class as a function");
            }(this, u2), e3 = c2.call(this), p(a(e3), t3), e3.opts = t3, e3.query = t3.query, e3.readyState = "", e3.socket = t3.socket, e3;
          }
          return e2 = u2, (n2 = [{ key: "onError", value: function(t3, e3) {
            var n3 = new Error(t3);
            return n3.type = "TransportError", n3.description = e3, this.emit("error", n3), this;
          } }, { key: "open", value: function() {
            return this.readyState !== "closed" && this.readyState !== "" || (this.readyState = "opening", this.doOpen()), this;
          } }, { key: "close", value: function() {
            return this.readyState !== "opening" && this.readyState !== "open" || (this.doClose(), this.onClose()), this;
          } }, { key: "send", value: function(t3) {
            this.readyState === "open" && this.write(t3);
          } }, { key: "onOpen", value: function() {
            this.readyState = "open", this.writable = true, this.emit("open");
          } }, { key: "onData", value: function(t3) {
            var e3 = f.decodePacket(t3, this.socket.binaryType);
            this.onPacket(e3);
          } }, { key: "onPacket", value: function(t3) {
            this.emit("packet", t3);
          } }, { key: "onClose", value: function() {
            this.readyState = "closed", this.emit("close");
          } }]) && o(e2.prototype, n2), r2 && o(e2, r2), u2;
        }(l);
        t.exports = h;
      }, function(t, e) {
        e.encode = function(t2) {
          var e2 = "";
          for (var n in t2)
            t2.hasOwnProperty(n) && (e2.length && (e2 += "&"), e2 += encodeURIComponent(n) + "=" + encodeURIComponent(t2[n]));
          return e2;
        }, e.decode = function(t2) {
          for (var e2 = {}, n = t2.split("&"), r = 0, o = n.length; r < o; r++) {
            var i = n[r].split("=");
            e2[decodeURIComponent(i[0])] = decodeURIComponent(i[1]);
          }
          return e2;
        };
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2, n2) {
          return (o = typeof Reflect != "undefined" && Reflect.get ? Reflect.get : function(t3, e3, n3) {
            var r2 = function(t4, e4) {
              for (; !Object.prototype.hasOwnProperty.call(t4, e4) && (t4 = a(t4)) !== null; )
                ;
              return t4;
            }(t3, e3);
            if (r2) {
              var o2 = Object.getOwnPropertyDescriptor(r2, e3);
              return o2.get ? o2.get.call(n3) : o2.value;
            }
          })(t2, e2, n2 || t2);
        }
        function i(t2, e2) {
          return (i = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function s(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = a(t2);
            if (e2) {
              var o2 = a(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return c(this, n2);
          };
        }
        function c(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return function(t3) {
            if (t3 === void 0)
              throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
            return t3;
          }(t2);
        }
        function a(t2) {
          return (a = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        function u(t2, e2) {
          if (!(t2 instanceof e2))
            throw new TypeError("Cannot call a class as a function");
        }
        function f(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function l(t2, e2, n2) {
          return e2 && f(t2.prototype, e2), n2 && f(t2, n2), t2;
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.Decoder = e.Encoder = e.PacketType = e.protocol = void 0;
        var p, h = n(2), y = n(30), d = n(15);
        e.protocol = 5, function(t2) {
          t2[t2.CONNECT = 0] = "CONNECT", t2[t2.DISCONNECT = 1] = "DISCONNECT", t2[t2.EVENT = 2] = "EVENT", t2[t2.ACK = 3] = "ACK", t2[t2.CONNECT_ERROR = 4] = "CONNECT_ERROR", t2[t2.BINARY_EVENT = 5] = "BINARY_EVENT", t2[t2.BINARY_ACK = 6] = "BINARY_ACK";
        }(p = e.PacketType || (e.PacketType = {}));
        var v = function() {
          function t2() {
            u(this, t2);
          }
          return l(t2, [{ key: "encode", value: function(t3) {
            return t3.type !== p.EVENT && t3.type !== p.ACK || !d.hasBinary(t3) ? [this.encodeAsString(t3)] : (t3.type = t3.type === p.EVENT ? p.BINARY_EVENT : p.BINARY_ACK, this.encodeAsBinary(t3));
          } }, { key: "encodeAsString", value: function(t3) {
            var e2 = "" + t3.type;
            return t3.type !== p.BINARY_EVENT && t3.type !== p.BINARY_ACK || (e2 += t3.attachments + "-"), t3.nsp && t3.nsp !== "/" && (e2 += t3.nsp + ","), t3.id != null && (e2 += t3.id), t3.data != null && (e2 += JSON.stringify(t3.data)), e2;
          } }, { key: "encodeAsBinary", value: function(t3) {
            var e2 = y.deconstructPacket(t3), n2 = this.encodeAsString(e2.packet), r2 = e2.buffers;
            return r2.unshift(n2), r2;
          } }]), t2;
        }();
        e.Encoder = v;
        var b = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && i(t3, e3);
          }(n2, t2);
          var e2 = s(n2);
          function n2() {
            return u(this, n2), e2.call(this);
          }
          return l(n2, [{ key: "add", value: function(t3) {
            var e3;
            if (typeof t3 == "string")
              (e3 = this.decodeString(t3)).type === p.BINARY_EVENT || e3.type === p.BINARY_ACK ? (this.reconstructor = new m(e3), e3.attachments === 0 && o(a(n2.prototype), "emit", this).call(this, "decoded", e3)) : o(a(n2.prototype), "emit", this).call(this, "decoded", e3);
            else {
              if (!d.isBinary(t3) && !t3.base64)
                throw new Error("Unknown type: " + t3);
              if (!this.reconstructor)
                throw new Error("got binary data when not reconstructing a packet");
              (e3 = this.reconstructor.takeBinaryData(t3)) && (this.reconstructor = null, o(a(n2.prototype), "emit", this).call(this, "decoded", e3));
            }
          } }, { key: "decodeString", value: function(t3) {
            var e3 = 0, r2 = { type: Number(t3.charAt(0)) };
            if (p[r2.type] === void 0)
              throw new Error("unknown packet type " + r2.type);
            if (r2.type === p.BINARY_EVENT || r2.type === p.BINARY_ACK) {
              for (var o2 = e3 + 1; t3.charAt(++e3) !== "-" && e3 != t3.length; )
                ;
              var i2 = t3.substring(o2, e3);
              if (i2 != Number(i2) || t3.charAt(e3) !== "-")
                throw new Error("Illegal attachments");
              r2.attachments = Number(i2);
            }
            if (t3.charAt(e3 + 1) === "/") {
              for (var s2 = e3 + 1; ++e3; ) {
                if (t3.charAt(e3) === ",")
                  break;
                if (e3 === t3.length)
                  break;
              }
              r2.nsp = t3.substring(s2, e3);
            } else
              r2.nsp = "/";
            var c2 = t3.charAt(e3 + 1);
            if (c2 !== "" && Number(c2) == c2) {
              for (var a2 = e3 + 1; ++e3; ) {
                var u2 = t3.charAt(e3);
                if (u2 == null || Number(u2) != u2) {
                  --e3;
                  break;
                }
                if (e3 === t3.length)
                  break;
              }
              r2.id = Number(t3.substring(a2, e3 + 1));
            }
            if (t3.charAt(++e3)) {
              var f2 = function(t4) {
                try {
                  return JSON.parse(t4);
                } catch (t5) {
                  return false;
                }
              }(t3.substr(e3));
              if (!n2.isPayloadValid(r2.type, f2))
                throw new Error("invalid payload");
              r2.data = f2;
            }
            return r2;
          } }, { key: "destroy", value: function() {
            this.reconstructor && this.reconstructor.finishedReconstruction();
          } }], [{ key: "isPayloadValid", value: function(t3, e3) {
            switch (t3) {
              case p.CONNECT:
                return r(e3) === "object";
              case p.DISCONNECT:
                return e3 === void 0;
              case p.CONNECT_ERROR:
                return typeof e3 == "string" || r(e3) === "object";
              case p.EVENT:
              case p.BINARY_EVENT:
                return Array.isArray(e3) && e3.length > 0;
              case p.ACK:
              case p.BINARY_ACK:
                return Array.isArray(e3);
            }
          } }]), n2;
        }(h);
        e.Decoder = b;
        var m = function() {
          function t2(e2) {
            u(this, t2), this.packet = e2, this.buffers = [], this.reconPack = e2;
          }
          return l(t2, [{ key: "takeBinaryData", value: function(t3) {
            if (this.buffers.push(t3), this.buffers.length === this.reconPack.attachments) {
              var e2 = y.reconstructPacket(this.reconPack, this.buffers);
              return this.finishedReconstruction(), e2;
            }
            return null;
          } }, { key: "finishedReconstruction", value: function() {
            this.reconPack = null, this.buffers = [];
          } }]), t2;
        }();
      }, function(t, e) {
        var n = /^(?:(?![^:@]+:[^:@\/]*@)(http|https|ws|wss):\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?((?:[a-f0-9]{0,4}:){2,7}[a-f0-9]{0,4}|[^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/, r = ["source", "protocol", "authority", "userInfo", "user", "password", "host", "port", "relative", "path", "directory", "file", "query", "anchor"];
        t.exports = function(t2) {
          var e2 = t2, o = t2.indexOf("["), i = t2.indexOf("]");
          o != -1 && i != -1 && (t2 = t2.substring(0, o) + t2.substring(o, i).replace(/:/g, ";") + t2.substring(i, t2.length));
          for (var s, c, a = n.exec(t2 || ""), u = {}, f = 14; f--; )
            u[r[f]] = a[f] || "";
          return o != -1 && i != -1 && (u.source = e2, u.host = u.host.substring(1, u.host.length - 1).replace(/;/g, ":"), u.authority = u.authority.replace("[", "").replace("]", "").replace(/;/g, ":"), u.ipv6uri = true), u.pathNames = function(t3, e3) {
            var n2 = e3.replace(/\/{2,9}/g, "/").split("/");
            e3.substr(0, 1) != "/" && e3.length !== 0 || n2.splice(0, 1);
            e3.substr(e3.length - 1, 1) == "/" && n2.splice(n2.length - 1, 1);
            return n2;
          }(0, u.path), u.queryKey = (s = u.query, c = {}, s.replace(/(?:^|&)([^&=]*)=?([^&]*)/g, function(t3, e3, n2) {
            e3 && (c[e3] = n2);
          }), c), u;
        };
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function i(t2, e2) {
          return (i = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function s(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = u(t2);
            if (e2) {
              var o2 = u(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return c(this, n2);
          };
        }
        function c(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return a(t2);
        }
        function a(t2) {
          if (t2 === void 0)
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
          return t2;
        }
        function u(t2) {
          return (u = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.Manager = void 0;
        var f = n(20), l = n(3), p = n(14), h = n(6), y = n(16), d = n(31), v = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && i(t3, e3);
          }(v2, t2);
          var e2, n2, c2, u2 = s(v2);
          function v2(t3, e3) {
            var n3, o2;
            !function(t4, e4) {
              if (!(t4 instanceof e4))
                throw new TypeError("Cannot call a class as a function");
            }(this, v2), (n3 = u2.call(this)).nsps = {}, n3.subs = [], t3 && r(t3) === "object" && (e3 = t3, t3 = void 0), (e3 = e3 || {}).path = e3.path || "/socket.io", n3.opts = e3, (0, l.installTimerFunctions)(a(n3), e3), n3.reconnection(e3.reconnection !== false), n3.reconnectionAttempts(e3.reconnectionAttempts || 1 / 0), n3.reconnectionDelay(e3.reconnectionDelay || 1e3), n3.reconnectionDelayMax(e3.reconnectionDelayMax || 5e3), n3.randomizationFactor((o2 = e3.randomizationFactor) !== null && o2 !== void 0 ? o2 : 0.5), n3.backoff = new d({ min: n3.reconnectionDelay(), max: n3.reconnectionDelayMax(), jitter: n3.randomizationFactor() }), n3.timeout(e3.timeout == null ? 2e4 : e3.timeout), n3._readyState = "closed", n3.uri = t3;
            var i2 = e3.parser || h;
            return n3.encoder = new i2.Encoder(), n3.decoder = new i2.Decoder(), n3._autoConnect = e3.autoConnect !== false, n3._autoConnect && n3.open(), n3;
          }
          return e2 = v2, (n2 = [{ key: "reconnection", value: function(t3) {
            return arguments.length ? (this._reconnection = !!t3, this) : this._reconnection;
          } }, { key: "reconnectionAttempts", value: function(t3) {
            return t3 === void 0 ? this._reconnectionAttempts : (this._reconnectionAttempts = t3, this);
          } }, { key: "reconnectionDelay", value: function(t3) {
            var e3;
            return t3 === void 0 ? this._reconnectionDelay : (this._reconnectionDelay = t3, (e3 = this.backoff) === null || e3 === void 0 || e3.setMin(t3), this);
          } }, { key: "randomizationFactor", value: function(t3) {
            var e3;
            return t3 === void 0 ? this._randomizationFactor : (this._randomizationFactor = t3, (e3 = this.backoff) === null || e3 === void 0 || e3.setJitter(t3), this);
          } }, { key: "reconnectionDelayMax", value: function(t3) {
            var e3;
            return t3 === void 0 ? this._reconnectionDelayMax : (this._reconnectionDelayMax = t3, (e3 = this.backoff) === null || e3 === void 0 || e3.setMax(t3), this);
          } }, { key: "timeout", value: function(t3) {
            return arguments.length ? (this._timeout = t3, this) : this._timeout;
          } }, { key: "maybeReconnectOnOpen", value: function() {
            !this._reconnecting && this._reconnection && this.backoff.attempts === 0 && this.reconnect();
          } }, { key: "open", value: function(t3) {
            var e3 = this;
            if (~this._readyState.indexOf("open"))
              return this;
            this.engine = f(this.uri, this.opts);
            var n3 = this.engine, r2 = this;
            this._readyState = "opening", this.skipReconnect = false;
            var o2 = (0, y.on)(n3, "open", function() {
              r2.onopen(), t3 && t3();
            }), i2 = (0, y.on)(n3, "error", function(n4) {
              r2.cleanup(), r2._readyState = "closed", e3.emitReserved("error", n4), t3 ? t3(n4) : r2.maybeReconnectOnOpen();
            });
            if (this._timeout !== false) {
              var s2 = this._timeout;
              s2 === 0 && o2();
              var c3 = this.setTimeoutFn(function() {
                o2(), n3.close(), n3.emit("error", new Error("timeout"));
              }, s2);
              this.opts.autoUnref && c3.unref(), this.subs.push(function() {
                clearTimeout(c3);
              });
            }
            return this.subs.push(o2), this.subs.push(i2), this;
          } }, { key: "connect", value: function(t3) {
            return this.open(t3);
          } }, { key: "onopen", value: function() {
            this.cleanup(), this._readyState = "open", this.emitReserved("open");
            var t3 = this.engine;
            this.subs.push((0, y.on)(t3, "ping", this.onping.bind(this)), (0, y.on)(t3, "data", this.ondata.bind(this)), (0, y.on)(t3, "error", this.onerror.bind(this)), (0, y.on)(t3, "close", this.onclose.bind(this)), (0, y.on)(this.decoder, "decoded", this.ondecoded.bind(this)));
          } }, { key: "onping", value: function() {
            this.emitReserved("ping");
          } }, { key: "ondata", value: function(t3) {
            this.decoder.add(t3);
          } }, { key: "ondecoded", value: function(t3) {
            this.emitReserved("packet", t3);
          } }, { key: "onerror", value: function(t3) {
            this.emitReserved("error", t3);
          } }, { key: "socket", value: function(t3, e3) {
            var n3 = this.nsps[t3];
            return n3 || (n3 = new p.Socket(this, t3, e3), this.nsps[t3] = n3), n3;
          } }, { key: "_destroy", value: function(t3) {
            for (var e3 = 0, n3 = Object.keys(this.nsps); e3 < n3.length; e3++) {
              var r2 = n3[e3];
              if (this.nsps[r2].active)
                return;
            }
            this._close();
          } }, { key: "_packet", value: function(t3) {
            for (var e3 = this.encoder.encode(t3), n3 = 0; n3 < e3.length; n3++)
              this.engine.write(e3[n3], t3.options);
          } }, { key: "cleanup", value: function() {
            this.subs.forEach(function(t3) {
              return t3();
            }), this.subs.length = 0, this.decoder.destroy();
          } }, { key: "_close", value: function() {
            this.skipReconnect = true, this._reconnecting = false, this._readyState === "opening" && this.cleanup(), this.backoff.reset(), this._readyState = "closed", this.engine && this.engine.close();
          } }, { key: "disconnect", value: function() {
            return this._close();
          } }, { key: "onclose", value: function(t3) {
            this.cleanup(), this.backoff.reset(), this._readyState = "closed", this.emitReserved("close", t3), this._reconnection && !this.skipReconnect && this.reconnect();
          } }, { key: "reconnect", value: function() {
            var t3 = this;
            if (this._reconnecting || this.skipReconnect)
              return this;
            var e3 = this;
            if (this.backoff.attempts >= this._reconnectionAttempts)
              this.backoff.reset(), this.emitReserved("reconnect_failed"), this._reconnecting = false;
            else {
              var n3 = this.backoff.duration();
              this._reconnecting = true;
              var r2 = this.setTimeoutFn(function() {
                e3.skipReconnect || (t3.emitReserved("reconnect_attempt", e3.backoff.attempts), e3.skipReconnect || e3.open(function(n4) {
                  n4 ? (e3._reconnecting = false, e3.reconnect(), t3.emitReserved("reconnect_error", n4)) : e3.onreconnect();
                }));
              }, n3);
              this.opts.autoUnref && r2.unref(), this.subs.push(function() {
                clearTimeout(r2);
              });
            }
          } }, { key: "onreconnect", value: function() {
            var t3 = this.backoff.attempts;
            this._reconnecting = false, this.backoff.reset(), this.emitReserved("reconnect", t3);
          } }]) && o(e2.prototype, n2), c2 && o(e2, c2), v2;
        }(n(17).StrictEventEmitter);
        e.Manager = v;
      }, function(t, e, n) {
        var r = n(10), o = n(23), i = n(27), s = n(28);
        e.polling = function(t2) {
          var e2 = false, n2 = false, s2 = t2.jsonp !== false;
          if (typeof location != "undefined") {
            var c = location.protocol === "https:", a = location.port;
            a || (a = c ? 443 : 80), e2 = t2.hostname !== location.hostname || a !== t2.port, n2 = t2.secure !== c;
          }
          if (t2.xdomain = e2, t2.xscheme = n2, "open" in new r(t2) && !t2.forceJSONP)
            return new o(t2);
          if (!s2)
            throw new Error("JSONP disabled");
          return new i(t2);
        }, e.websocket = s;
      }, function(t, e, n) {
        var r = n(22), o = n(0);
        t.exports = function(t2) {
          var e2 = t2.xdomain, n2 = t2.xscheme, i = t2.enablesXDR;
          try {
            if (typeof XMLHttpRequest != "undefined" && (!e2 || r))
              return new XMLHttpRequest();
          } catch (t3) {
          }
          try {
            if (typeof XDomainRequest != "undefined" && !n2 && i)
              return new XDomainRequest();
          } catch (t3) {
          }
          if (!e2)
            try {
              return new o[["Active"].concat("Object").join("X")]("Microsoft.XMLHTTP");
            } catch (t3) {
            }
        };
      }, function(t, e, n) {
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          if (!(t2 instanceof e2))
            throw new TypeError("Cannot call a class as a function");
        }
        function i(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function s(t2, e2) {
          return (s = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function c(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = u(t2);
            if (e2) {
              var o2 = u(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return a(this, n2);
          };
        }
        function a(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return function(t3) {
            if (t3 === void 0)
              throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
            return t3;
          }(t2);
        }
        function u(t2) {
          return (u = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var f = n(4), l = n(5), p = n(1), h = n(13), y = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && s(t3, e3);
          }(u2, t2);
          var e2, n2, r2, a2 = c(u2);
          function u2() {
            return o(this, u2), a2.apply(this, arguments);
          }
          return e2 = u2, (n2 = [{ key: "name", get: function() {
            return "polling";
          } }, { key: "doOpen", value: function() {
            this.poll();
          } }, { key: "pause", value: function(t3) {
            var e3 = this;
            this.readyState = "pausing";
            var n3 = function() {
              e3.readyState = "paused", t3();
            };
            if (this.polling || !this.writable) {
              var r3 = 0;
              this.polling && (r3++, this.once("pollComplete", function() {
                --r3 || n3();
              })), this.writable || (r3++, this.once("drain", function() {
                --r3 || n3();
              }));
            } else
              n3();
          } }, { key: "poll", value: function() {
            this.polling = true, this.doPoll(), this.emit("poll");
          } }, { key: "onData", value: function(t3) {
            var e3 = this;
            p.decodePayload(t3, this.socket.binaryType).forEach(function(t4) {
              if (e3.readyState === "opening" && t4.type === "open" && e3.onOpen(), t4.type === "close")
                return e3.onClose(), false;
              e3.onPacket(t4);
            }), this.readyState !== "closed" && (this.polling = false, this.emit("pollComplete"), this.readyState === "open" && this.poll());
          } }, { key: "doClose", value: function() {
            var t3 = this, e3 = function() {
              t3.write([{ type: "close" }]);
            };
            this.readyState === "open" ? e3() : this.once("open", e3);
          } }, { key: "write", value: function(t3) {
            var e3 = this;
            this.writable = false, p.encodePayload(t3, function(t4) {
              e3.doWrite(t4, function() {
                e3.writable = true, e3.emit("drain");
              });
            });
          } }, { key: "uri", value: function() {
            var t3 = this.query || {}, e3 = this.opts.secure ? "https" : "http", n3 = "";
            return this.opts.timestampRequests !== false && (t3[this.opts.timestampParam] = h()), this.supportsBinary || t3.sid || (t3.b64 = 1), t3 = l.encode(t3), this.opts.port && (e3 === "https" && Number(this.opts.port) !== 443 || e3 === "http" && Number(this.opts.port) !== 80) && (n3 = ":" + this.opts.port), t3.length && (t3 = "?" + t3), e3 + "://" + (this.opts.hostname.indexOf(":") !== -1 ? "[" + this.opts.hostname + "]" : this.opts.hostname) + n3 + this.opts.path + t3;
          } }]) && i(e2.prototype, n2), r2 && i(e2, r2), u2;
        }(f);
        t.exports = y;
      }, function(t, e) {
        var n = Object.create(null);
        n.open = "0", n.close = "1", n.ping = "2", n.pong = "3", n.message = "4", n.upgrade = "5", n.noop = "6";
        var r = Object.create(null);
        Object.keys(n).forEach(function(t2) {
          r[n[t2]] = t2;
        });
        t.exports = { PACKET_TYPES: n, PACKET_TYPES_REVERSE: r, ERROR_PACKET: { type: "error", data: "parser error" } };
      }, function(t, e, n) {
        "use strict";
        var r, o = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_".split(""), i = {}, s = 0, c = 0;
        function a(t2) {
          var e2 = "";
          do {
            e2 = o[t2 % 64] + e2, t2 = Math.floor(t2 / 64);
          } while (t2 > 0);
          return e2;
        }
        function u() {
          var t2 = a(+new Date());
          return t2 !== r ? (s = 0, r = t2) : t2 + "." + a(s++);
        }
        for (; c < 64; c++)
          i[o[c]] = c;
        u.encode = a, u.decode = function(t2) {
          var e2 = 0;
          for (c = 0; c < t2.length; c++)
            e2 = 64 * e2 + i[t2.charAt(c)];
          return e2;
        }, t.exports = u;
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          var n2 = typeof Symbol != "undefined" && t2[Symbol.iterator] || t2["@@iterator"];
          if (!n2) {
            if (Array.isArray(t2) || (n2 = function(t3, e3) {
              if (!t3)
                return;
              if (typeof t3 == "string")
                return i(t3, e3);
              var n3 = Object.prototype.toString.call(t3).slice(8, -1);
              n3 === "Object" && t3.constructor && (n3 = t3.constructor.name);
              if (n3 === "Map" || n3 === "Set")
                return Array.from(t3);
              if (n3 === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n3))
                return i(t3, e3);
            }(t2)) || e2 && t2 && typeof t2.length == "number") {
              n2 && (t2 = n2);
              var r2 = 0, o2 = function() {
              };
              return { s: o2, n: function() {
                return r2 >= t2.length ? { done: true } : { done: false, value: t2[r2++] };
              }, e: function(t3) {
                throw t3;
              }, f: o2 };
            }
            throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
          }
          var s2, c2 = true, a2 = false;
          return { s: function() {
            n2 = n2.call(t2);
          }, n: function() {
            var t3 = n2.next();
            return c2 = t3.done, t3;
          }, e: function(t3) {
            a2 = true, s2 = t3;
          }, f: function() {
            try {
              c2 || n2.return == null || n2.return();
            } finally {
              if (a2)
                throw s2;
            }
          } };
        }
        function i(t2, e2) {
          (e2 == null || e2 > t2.length) && (e2 = t2.length);
          for (var n2 = 0, r2 = new Array(e2); n2 < e2; n2++)
            r2[n2] = t2[n2];
          return r2;
        }
        function s(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function c(t2, e2, n2) {
          return (c = typeof Reflect != "undefined" && Reflect.get ? Reflect.get : function(t3, e3, n3) {
            var r2 = function(t4, e4) {
              for (; !Object.prototype.hasOwnProperty.call(t4, e4) && (t4 = l(t4)) !== null; )
                ;
              return t4;
            }(t3, e3);
            if (r2) {
              var o2 = Object.getOwnPropertyDescriptor(r2, e3);
              return o2.get ? o2.get.call(n3) : o2.value;
            }
          })(t2, e2, n2 || t2);
        }
        function a(t2, e2) {
          return (a = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function u(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = l(t2);
            if (e2) {
              var o2 = l(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return f(this, n2);
          };
        }
        function f(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return function(t3) {
            if (t3 === void 0)
              throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
            return t3;
          }(t2);
        }
        function l(t2) {
          return (l = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.Socket = void 0;
        var p = n(6), h = n(16), y = n(17), d = Object.freeze({ connect: 1, connect_error: 1, disconnect: 1, disconnecting: 1, newListener: 1, removeListener: 1 }), v = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && a(t3, e3);
          }(f2, t2);
          var e2, n2, r2, i2 = u(f2);
          function f2(t3, e3, n3) {
            var r3;
            return function(t4, e4) {
              if (!(t4 instanceof e4))
                throw new TypeError("Cannot call a class as a function");
            }(this, f2), (r3 = i2.call(this)).connected = false, r3.disconnected = true, r3.receiveBuffer = [], r3.sendBuffer = [], r3.ids = 0, r3.acks = {}, r3.flags = {}, r3.io = t3, r3.nsp = e3, n3 && n3.auth && (r3.auth = n3.auth), r3.io._autoConnect && r3.open(), r3;
          }
          return e2 = f2, (n2 = [{ key: "subEvents", value: function() {
            if (!this.subs) {
              var t3 = this.io;
              this.subs = [(0, h.on)(t3, "open", this.onopen.bind(this)), (0, h.on)(t3, "packet", this.onpacket.bind(this)), (0, h.on)(t3, "error", this.onerror.bind(this)), (0, h.on)(t3, "close", this.onclose.bind(this))];
            }
          } }, { key: "active", get: function() {
            return !!this.subs;
          } }, { key: "connect", value: function() {
            return this.connected || (this.subEvents(), this.io._reconnecting || this.io.open(), this.io._readyState === "open" && this.onopen()), this;
          } }, { key: "open", value: function() {
            return this.connect();
          } }, { key: "send", value: function() {
            for (var t3 = arguments.length, e3 = new Array(t3), n3 = 0; n3 < t3; n3++)
              e3[n3] = arguments[n3];
            return e3.unshift("message"), this.emit.apply(this, e3), this;
          } }, { key: "emit", value: function(t3) {
            if (d.hasOwnProperty(t3))
              throw new Error('"' + t3 + '" is a reserved event name');
            for (var e3 = arguments.length, n3 = new Array(e3 > 1 ? e3 - 1 : 0), r3 = 1; r3 < e3; r3++)
              n3[r3 - 1] = arguments[r3];
            n3.unshift(t3);
            var o2 = { type: p.PacketType.EVENT, data: n3, options: {} };
            o2.options.compress = this.flags.compress !== false, typeof n3[n3.length - 1] == "function" && (this.acks[this.ids] = n3.pop(), o2.id = this.ids++);
            var i3 = this.io.engine && this.io.engine.transport && this.io.engine.transport.writable, s2 = this.flags.volatile && (!i3 || !this.connected);
            return s2 || (this.connected ? this.packet(o2) : this.sendBuffer.push(o2)), this.flags = {}, this;
          } }, { key: "packet", value: function(t3) {
            t3.nsp = this.nsp, this.io._packet(t3);
          } }, { key: "onopen", value: function() {
            var t3 = this;
            typeof this.auth == "function" ? this.auth(function(e3) {
              t3.packet({ type: p.PacketType.CONNECT, data: e3 });
            }) : this.packet({ type: p.PacketType.CONNECT, data: this.auth });
          } }, { key: "onerror", value: function(t3) {
            this.connected || this.emitReserved("connect_error", t3);
          } }, { key: "onclose", value: function(t3) {
            this.connected = false, this.disconnected = true, delete this.id, this.emitReserved("disconnect", t3);
          } }, { key: "onpacket", value: function(t3) {
            if (t3.nsp === this.nsp)
              switch (t3.type) {
                case p.PacketType.CONNECT:
                  if (t3.data && t3.data.sid) {
                    var e3 = t3.data.sid;
                    this.onconnect(e3);
                  } else
                    this.emitReserved("connect_error", new Error("It seems you are trying to reach a Socket.IO server in v2.x with a v3.x client, but they are not compatible (more information here: https://socket.io/docs/v3/migrating-from-2-x-to-3-0/)"));
                  break;
                case p.PacketType.EVENT:
                case p.PacketType.BINARY_EVENT:
                  this.onevent(t3);
                  break;
                case p.PacketType.ACK:
                case p.PacketType.BINARY_ACK:
                  this.onack(t3);
                  break;
                case p.PacketType.DISCONNECT:
                  this.ondisconnect();
                  break;
                case p.PacketType.CONNECT_ERROR:
                  var n3 = new Error(t3.data.message);
                  n3.data = t3.data.data, this.emitReserved("connect_error", n3);
              }
          } }, { key: "onevent", value: function(t3) {
            var e3 = t3.data || [];
            t3.id != null && e3.push(this.ack(t3.id)), this.connected ? this.emitEvent(e3) : this.receiveBuffer.push(Object.freeze(e3));
          } }, { key: "emitEvent", value: function(t3) {
            if (this._anyListeners && this._anyListeners.length) {
              var e3, n3 = o(this._anyListeners.slice());
              try {
                for (n3.s(); !(e3 = n3.n()).done; )
                  e3.value.apply(this, t3);
              } catch (t4) {
                n3.e(t4);
              } finally {
                n3.f();
              }
            }
            c(l(f2.prototype), "emit", this).apply(this, t3);
          } }, { key: "ack", value: function(t3) {
            var e3 = this, n3 = false;
            return function() {
              if (!n3) {
                n3 = true;
                for (var r3 = arguments.length, o2 = new Array(r3), i3 = 0; i3 < r3; i3++)
                  o2[i3] = arguments[i3];
                e3.packet({ type: p.PacketType.ACK, id: t3, data: o2 });
              }
            };
          } }, { key: "onack", value: function(t3) {
            var e3 = this.acks[t3.id];
            typeof e3 == "function" && (e3.apply(this, t3.data), delete this.acks[t3.id]);
          } }, { key: "onconnect", value: function(t3) {
            this.id = t3, this.connected = true, this.disconnected = false, this.emitBuffered(), this.emitReserved("connect");
          } }, { key: "emitBuffered", value: function() {
            var t3 = this;
            this.receiveBuffer.forEach(function(e3) {
              return t3.emitEvent(e3);
            }), this.receiveBuffer = [], this.sendBuffer.forEach(function(e3) {
              return t3.packet(e3);
            }), this.sendBuffer = [];
          } }, { key: "ondisconnect", value: function() {
            this.destroy(), this.onclose("io server disconnect");
          } }, { key: "destroy", value: function() {
            this.subs && (this.subs.forEach(function(t3) {
              return t3();
            }), this.subs = void 0), this.io._destroy(this);
          } }, { key: "disconnect", value: function() {
            return this.connected && this.packet({ type: p.PacketType.DISCONNECT }), this.destroy(), this.connected && this.onclose("io client disconnect"), this;
          } }, { key: "close", value: function() {
            return this.disconnect();
          } }, { key: "compress", value: function(t3) {
            return this.flags.compress = t3, this;
          } }, { key: "volatile", get: function() {
            return this.flags.volatile = true, this;
          } }, { key: "onAny", value: function(t3) {
            return this._anyListeners = this._anyListeners || [], this._anyListeners.push(t3), this;
          } }, { key: "prependAny", value: function(t3) {
            return this._anyListeners = this._anyListeners || [], this._anyListeners.unshift(t3), this;
          } }, { key: "offAny", value: function(t3) {
            if (!this._anyListeners)
              return this;
            if (t3) {
              for (var e3 = this._anyListeners, n3 = 0; n3 < e3.length; n3++)
                if (t3 === e3[n3])
                  return e3.splice(n3, 1), this;
            } else
              this._anyListeners = [];
            return this;
          } }, { key: "listenersAny", value: function() {
            return this._anyListeners || [];
          } }]) && s(e2.prototype, n2), r2 && s(e2, r2), f2;
        }(y.StrictEventEmitter);
        e.Socket = v;
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.hasBinary = e.isBinary = void 0;
        var o = typeof ArrayBuffer == "function", i = Object.prototype.toString, s = typeof Blob == "function" || typeof Blob != "undefined" && i.call(Blob) === "[object BlobConstructor]", c = typeof File == "function" || typeof File != "undefined" && i.call(File) === "[object FileConstructor]";
        function a(t2) {
          return o && (t2 instanceof ArrayBuffer || function(t3) {
            return typeof ArrayBuffer.isView == "function" ? ArrayBuffer.isView(t3) : t3.buffer instanceof ArrayBuffer;
          }(t2)) || s && t2 instanceof Blob || c && t2 instanceof File;
        }
        e.isBinary = a, e.hasBinary = function t2(e2, n2) {
          if (!e2 || r(e2) !== "object")
            return false;
          if (Array.isArray(e2)) {
            for (var o2 = 0, i2 = e2.length; o2 < i2; o2++)
              if (t2(e2[o2]))
                return true;
            return false;
          }
          if (a(e2))
            return true;
          if (e2.toJSON && typeof e2.toJSON == "function" && arguments.length === 1)
            return t2(e2.toJSON(), true);
          for (var s2 in e2)
            if (Object.prototype.hasOwnProperty.call(e2, s2) && t2(e2[s2]))
              return true;
          return false;
        };
      }, function(t, e, n) {
        "use strict";
        Object.defineProperty(e, "__esModule", { value: true }), e.on = void 0, e.on = function(t2, e2, n2) {
          return t2.on(e2, n2), function() {
            t2.off(e2, n2);
          };
        };
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          if (!(t2 instanceof e2))
            throw new TypeError("Cannot call a class as a function");
        }
        function i(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function s(t2, e2, n2) {
          return (s = typeof Reflect != "undefined" && Reflect.get ? Reflect.get : function(t3, e3, n3) {
            var r2 = function(t4, e4) {
              for (; !Object.prototype.hasOwnProperty.call(t4, e4) && (t4 = f(t4)) !== null; )
                ;
              return t4;
            }(t3, e3);
            if (r2) {
              var o2 = Object.getOwnPropertyDescriptor(r2, e3);
              return o2.get ? o2.get.call(n3) : o2.value;
            }
          })(t2, e2, n2 || t2);
        }
        function c(t2, e2) {
          return (c = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function a(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = f(t2);
            if (e2) {
              var o2 = f(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return u(this, n2);
          };
        }
        function u(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return function(t3) {
            if (t3 === void 0)
              throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
            return t3;
          }(t2);
        }
        function f(t2) {
          return (f = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.StrictEventEmitter = void 0;
        var l = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && c(t3, e3);
          }(l2, t2);
          var e2, n2, r2, u2 = a(l2);
          function l2() {
            return o(this, l2), u2.apply(this, arguments);
          }
          return e2 = l2, (n2 = [{ key: "on", value: function(t3, e3) {
            return s(f(l2.prototype), "on", this).call(this, t3, e3), this;
          } }, { key: "once", value: function(t3, e3) {
            return s(f(l2.prototype), "once", this).call(this, t3, e3), this;
          } }, { key: "emit", value: function(t3) {
            for (var e3, n3 = arguments.length, r3 = new Array(n3 > 1 ? n3 - 1 : 0), o2 = 1; o2 < n3; o2++)
              r3[o2 - 1] = arguments[o2];
            return (e3 = s(f(l2.prototype), "emit", this)).call.apply(e3, [this, t3].concat(r3)), this;
          } }, { key: "emitReserved", value: function(t3) {
            for (var e3, n3 = arguments.length, r3 = new Array(n3 > 1 ? n3 - 1 : 0), o2 = 1; o2 < n3; o2++)
              r3[o2 - 1] = arguments[o2];
            return (e3 = s(f(l2.prototype), "emit", this)).call.apply(e3, [this, t3].concat(r3)), this;
          } }, { key: "listeners", value: function(t3) {
            return s(f(l2.prototype), "listeners", this).call(this, t3);
          } }]) && i(e2.prototype, n2), r2 && i(e2, r2), l2;
        }(n(2));
        e.StrictEventEmitter = l;
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.io = e.Socket = e.Manager = e.protocol = void 0;
        var o = n(19), i = n(8);
        t.exports = e = c;
        var s = e.managers = {};
        function c(t2, e2) {
          r(t2) === "object" && (e2 = t2, t2 = void 0), e2 = e2 || {};
          var n2, c2 = (0, o.url)(t2, e2.path || "/socket.io"), a2 = c2.source, u2 = c2.id, f2 = c2.path, l = s[u2] && f2 in s[u2].nsps;
          return e2.forceNew || e2["force new connection"] || e2.multiplex === false || l ? n2 = new i.Manager(a2, e2) : (s[u2] || (s[u2] = new i.Manager(a2, e2)), n2 = s[u2]), c2.query && !e2.query && (e2.query = c2.queryKey), n2.socket(c2.path, e2);
        }
        e.io = c;
        var a = n(6);
        Object.defineProperty(e, "protocol", { enumerable: true, get: function() {
          return a.protocol;
        } }), e.connect = c;
        var u = n(8);
        Object.defineProperty(e, "Manager", { enumerable: true, get: function() {
          return u.Manager;
        } });
        var f = n(14);
        Object.defineProperty(e, "Socket", { enumerable: true, get: function() {
          return f.Socket;
        } }), e.default = c;
      }, function(t, e, n) {
        "use strict";
        Object.defineProperty(e, "__esModule", { value: true }), e.url = void 0;
        var r = n(7);
        e.url = function(t2) {
          var e2 = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : "", n2 = arguments.length > 2 ? arguments[2] : void 0, o = t2;
          n2 = n2 || typeof location != "undefined" && location, t2 == null && (t2 = n2.protocol + "//" + n2.host), typeof t2 == "string" && (t2.charAt(0) === "/" && (t2 = t2.charAt(1) === "/" ? n2.protocol + t2 : n2.host + t2), /^(https?|wss?):\/\//.test(t2) || (t2 = n2 !== void 0 ? n2.protocol + "//" + t2 : "https://" + t2), o = r(t2)), o.port || (/^(http|ws)$/.test(o.protocol) ? o.port = "80" : /^(http|ws)s$/.test(o.protocol) && (o.port = "443")), o.path = o.path || "/";
          var i = o.host.indexOf(":") !== -1, s = i ? "[" + o.host + "]" : o.host;
          return o.id = o.protocol + "://" + s + ":" + o.port + e2, o.href = o.protocol + "://" + s + (n2 && n2.port === o.port ? "" : ":" + o.port), o;
        };
      }, function(t, e, n) {
        var r = n(21);
        t.exports = function(t2, e2) {
          return new r(t2, e2);
        }, t.exports.Socket = r, t.exports.protocol = r.protocol, t.exports.Transport = n(4), t.exports.transports = n(9), t.exports.parser = n(1);
      }, function(t, e, n) {
        function r() {
          return (r = Object.assign || function(t2) {
            for (var e2 = 1; e2 < arguments.length; e2++) {
              var n2 = arguments[e2];
              for (var r2 in n2)
                Object.prototype.hasOwnProperty.call(n2, r2) && (t2[r2] = n2[r2]);
            }
            return t2;
          }).apply(this, arguments);
        }
        function o(t2) {
          return (o = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function i(t2, e2) {
          if (!(t2 instanceof e2))
            throw new TypeError("Cannot call a class as a function");
        }
        function s(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function c(t2, e2) {
          return (c = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function a(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = l(t2);
            if (e2) {
              var o2 = l(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return u(this, n2);
          };
        }
        function u(t2, e2) {
          if (e2 && (o(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return f(t2);
        }
        function f(t2) {
          if (t2 === void 0)
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
          return t2;
        }
        function l(t2) {
          return (l = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var p = n(9), h = n(2), y = n(1), d = n(7), v = n(5), b = n(3).installTimerFunctions, m = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && c(t3, e3);
          }(h2, t2);
          var e2, n2, u2, l2 = a(h2);
          function h2(t3) {
            var e3, n3 = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
            return i(this, h2), e3 = l2.call(this), t3 && o(t3) === "object" && (n3 = t3, t3 = null), t3 ? (t3 = d(t3), n3.hostname = t3.host, n3.secure = t3.protocol === "https" || t3.protocol === "wss", n3.port = t3.port, t3.query && (n3.query = t3.query)) : n3.host && (n3.hostname = d(n3.host).host), b(f(e3), n3), e3.secure = n3.secure != null ? n3.secure : typeof location != "undefined" && location.protocol === "https:", n3.hostname && !n3.port && (n3.port = e3.secure ? "443" : "80"), e3.hostname = n3.hostname || (typeof location != "undefined" ? location.hostname : "localhost"), e3.port = n3.port || (typeof location != "undefined" && location.port ? location.port : e3.secure ? 443 : 80), e3.transports = n3.transports || ["polling", "websocket"], e3.readyState = "", e3.writeBuffer = [], e3.prevBufferLen = 0, e3.opts = r({ path: "/engine.io", agent: false, withCredentials: false, upgrade: true, jsonp: true, timestampParam: "t", rememberUpgrade: false, rejectUnauthorized: true, perMessageDeflate: { threshold: 1024 }, transportOptions: {}, closeOnBeforeunload: true }, n3), e3.opts.path = e3.opts.path.replace(/\/$/, "") + "/", typeof e3.opts.query == "string" && (e3.opts.query = v.decode(e3.opts.query)), e3.id = null, e3.upgrades = null, e3.pingInterval = null, e3.pingTimeout = null, e3.pingTimeoutTimer = null, typeof addEventListener == "function" && (e3.opts.closeOnBeforeunload && addEventListener("beforeunload", function() {
              e3.transport && (e3.transport.removeAllListeners(), e3.transport.close());
            }, false), e3.hostname !== "localhost" && (e3.offlineEventListener = function() {
              e3.onClose("transport close");
            }, addEventListener("offline", e3.offlineEventListener, false))), e3.open(), e3;
          }
          return e2 = h2, (n2 = [{ key: "createTransport", value: function(t3) {
            var e3 = function(t4) {
              var e4 = {};
              for (var n4 in t4)
                t4.hasOwnProperty(n4) && (e4[n4] = t4[n4]);
              return e4;
            }(this.opts.query);
            e3.EIO = y.protocol, e3.transport = t3, this.id && (e3.sid = this.id);
            var n3 = r({}, this.opts.transportOptions[t3], this.opts, { query: e3, socket: this, hostname: this.hostname, secure: this.secure, port: this.port });
            return new p[t3](n3);
          } }, { key: "open", value: function() {
            var t3, e3 = this;
            if (this.opts.rememberUpgrade && h2.priorWebsocketSuccess && this.transports.indexOf("websocket") !== -1)
              t3 = "websocket";
            else {
              if (this.transports.length === 0)
                return void this.setTimeoutFn(function() {
                  e3.emit("error", "No transports available");
                }, 0);
              t3 = this.transports[0];
            }
            this.readyState = "opening";
            try {
              t3 = this.createTransport(t3);
            } catch (t4) {
              return this.transports.shift(), void this.open();
            }
            t3.open(), this.setTransport(t3);
          } }, { key: "setTransport", value: function(t3) {
            var e3 = this;
            this.transport && this.transport.removeAllListeners(), this.transport = t3, t3.on("drain", this.onDrain.bind(this)).on("packet", this.onPacket.bind(this)).on("error", this.onError.bind(this)).on("close", function() {
              e3.onClose("transport close");
            });
          } }, { key: "probe", value: function(t3) {
            var e3 = this, n3 = this.createTransport(t3, { probe: 1 }), r2 = false;
            h2.priorWebsocketSuccess = false;
            var o2 = function() {
              r2 || (n3.send([{ type: "ping", data: "probe" }]), n3.once("packet", function(t4) {
                if (!r2)
                  if (t4.type === "pong" && t4.data === "probe") {
                    if (e3.upgrading = true, e3.emit("upgrading", n3), !n3)
                      return;
                    h2.priorWebsocketSuccess = n3.name === "websocket", e3.transport.pause(function() {
                      r2 || e3.readyState !== "closed" && (f2(), e3.setTransport(n3), n3.send([{ type: "upgrade" }]), e3.emit("upgrade", n3), n3 = null, e3.upgrading = false, e3.flush());
                    });
                  } else {
                    var o3 = new Error("probe error");
                    o3.transport = n3.name, e3.emit("upgradeError", o3);
                  }
              }));
            };
            function i2() {
              r2 || (r2 = true, f2(), n3.close(), n3 = null);
            }
            var s2 = function(t4) {
              var r3 = new Error("probe error: " + t4);
              r3.transport = n3.name, i2(), e3.emit("upgradeError", r3);
            };
            function c2() {
              s2("transport closed");
            }
            function a2() {
              s2("socket closed");
            }
            function u3(t4) {
              n3 && t4.name !== n3.name && i2();
            }
            var f2 = function() {
              n3.removeListener("open", o2), n3.removeListener("error", s2), n3.removeListener("close", c2), e3.removeListener("close", a2), e3.removeListener("upgrading", u3);
            };
            n3.once("open", o2), n3.once("error", s2), n3.once("close", c2), this.once("close", a2), this.once("upgrading", u3), n3.open();
          } }, { key: "onOpen", value: function() {
            if (this.readyState = "open", h2.priorWebsocketSuccess = this.transport.name === "websocket", this.emit("open"), this.flush(), this.readyState === "open" && this.opts.upgrade && this.transport.pause)
              for (var t3 = 0, e3 = this.upgrades.length; t3 < e3; t3++)
                this.probe(this.upgrades[t3]);
          } }, { key: "onPacket", value: function(t3) {
            if (this.readyState === "opening" || this.readyState === "open" || this.readyState === "closing")
              switch (this.emit("packet", t3), this.emit("heartbeat"), t3.type) {
                case "open":
                  this.onHandshake(JSON.parse(t3.data));
                  break;
                case "ping":
                  this.resetPingTimeout(), this.sendPacket("pong"), this.emit("ping"), this.emit("pong");
                  break;
                case "error":
                  var e3 = new Error("server error");
                  e3.code = t3.data, this.onError(e3);
                  break;
                case "message":
                  this.emit("data", t3.data), this.emit("message", t3.data);
              }
          } }, { key: "onHandshake", value: function(t3) {
            this.emit("handshake", t3), this.id = t3.sid, this.transport.query.sid = t3.sid, this.upgrades = this.filterUpgrades(t3.upgrades), this.pingInterval = t3.pingInterval, this.pingTimeout = t3.pingTimeout, this.onOpen(), this.readyState !== "closed" && this.resetPingTimeout();
          } }, { key: "resetPingTimeout", value: function() {
            var t3 = this;
            this.clearTimeoutFn(this.pingTimeoutTimer), this.pingTimeoutTimer = this.setTimeoutFn(function() {
              t3.onClose("ping timeout");
            }, this.pingInterval + this.pingTimeout), this.opts.autoUnref && this.pingTimeoutTimer.unref();
          } }, { key: "onDrain", value: function() {
            this.writeBuffer.splice(0, this.prevBufferLen), this.prevBufferLen = 0, this.writeBuffer.length === 0 ? this.emit("drain") : this.flush();
          } }, { key: "flush", value: function() {
            this.readyState !== "closed" && this.transport.writable && !this.upgrading && this.writeBuffer.length && (this.transport.send(this.writeBuffer), this.prevBufferLen = this.writeBuffer.length, this.emit("flush"));
          } }, { key: "write", value: function(t3, e3, n3) {
            return this.sendPacket("message", t3, e3, n3), this;
          } }, { key: "send", value: function(t3, e3, n3) {
            return this.sendPacket("message", t3, e3, n3), this;
          } }, { key: "sendPacket", value: function(t3, e3, n3, r2) {
            if (typeof e3 == "function" && (r2 = e3, e3 = void 0), typeof n3 == "function" && (r2 = n3, n3 = null), this.readyState !== "closing" && this.readyState !== "closed") {
              (n3 = n3 || {}).compress = n3.compress !== false;
              var o2 = { type: t3, data: e3, options: n3 };
              this.emit("packetCreate", o2), this.writeBuffer.push(o2), r2 && this.once("flush", r2), this.flush();
            }
          } }, { key: "close", value: function() {
            var t3 = this, e3 = function() {
              t3.onClose("forced close"), t3.transport.close();
            }, n3 = function n4() {
              t3.removeListener("upgrade", n4), t3.removeListener("upgradeError", n4), e3();
            }, r2 = function() {
              t3.once("upgrade", n3), t3.once("upgradeError", n3);
            };
            return this.readyState !== "opening" && this.readyState !== "open" || (this.readyState = "closing", this.writeBuffer.length ? this.once("drain", function() {
              t3.upgrading ? r2() : e3();
            }) : this.upgrading ? r2() : e3()), this;
          } }, { key: "onError", value: function(t3) {
            h2.priorWebsocketSuccess = false, this.emit("error", t3), this.onClose("transport error", t3);
          } }, { key: "onClose", value: function(t3, e3) {
            this.readyState !== "opening" && this.readyState !== "open" && this.readyState !== "closing" || (this.clearTimeoutFn(this.pingIntervalTimer), this.clearTimeoutFn(this.pingTimeoutTimer), this.transport.removeAllListeners("close"), this.transport.close(), this.transport.removeAllListeners(), typeof removeEventListener == "function" && removeEventListener("offline", this.offlineEventListener, false), this.readyState = "closed", this.id = null, this.emit("close", t3, e3), this.writeBuffer = [], this.prevBufferLen = 0);
          } }, { key: "filterUpgrades", value: function(t3) {
            for (var e3 = [], n3 = 0, r2 = t3.length; n3 < r2; n3++)
              ~this.transports.indexOf(t3[n3]) && e3.push(t3[n3]);
            return e3;
          } }]) && s(e2.prototype, n2), u2 && s(e2, u2), h2;
        }(h);
        m.priorWebsocketSuccess = false, m.protocol = y.protocol, t.exports = m;
      }, function(t, e) {
        try {
          t.exports = typeof XMLHttpRequest != "undefined" && "withCredentials" in new XMLHttpRequest();
        } catch (e2) {
          t.exports = false;
        }
      }, function(t, e, n) {
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o() {
          return (o = Object.assign || function(t2) {
            for (var e2 = 1; e2 < arguments.length; e2++) {
              var n2 = arguments[e2];
              for (var r2 in n2)
                Object.prototype.hasOwnProperty.call(n2, r2) && (t2[r2] = n2[r2]);
            }
            return t2;
          }).apply(this, arguments);
        }
        function i(t2, e2) {
          if (!(t2 instanceof e2))
            throw new TypeError("Cannot call a class as a function");
        }
        function s(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function c(t2, e2, n2) {
          return e2 && s(t2.prototype, e2), n2 && s(t2, n2), t2;
        }
        function a(t2, e2) {
          if (typeof e2 != "function" && e2 !== null)
            throw new TypeError("Super expression must either be null or a function");
          t2.prototype = Object.create(e2 && e2.prototype, { constructor: { value: t2, writable: true, configurable: true } }), e2 && u(t2, e2);
        }
        function u(t2, e2) {
          return (u = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function f(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = h(t2);
            if (e2) {
              var o2 = h(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return l(this, n2);
          };
        }
        function l(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return p(t2);
        }
        function p(t2) {
          if (t2 === void 0)
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
          return t2;
        }
        function h(t2) {
          return (h = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var y = n(10), d = n(11), v = n(2), b = n(3), m = b.pick, g = b.installTimerFunctions, k = n(0);
        function w() {
        }
        var O = new y({ xdomain: false }).responseType != null, _ = function(t2) {
          a(n2, t2);
          var e2 = f(n2);
          function n2(t3) {
            var r2;
            if (i(this, n2), r2 = e2.call(this, t3), typeof location != "undefined") {
              var o2 = location.protocol === "https:", s2 = location.port;
              s2 || (s2 = o2 ? 443 : 80), r2.xd = typeof location != "undefined" && t3.hostname !== location.hostname || s2 !== t3.port, r2.xs = t3.secure !== o2;
            }
            var c2 = t3 && t3.forceBase64;
            return r2.supportsBinary = O && !c2, r2;
          }
          return c(n2, [{ key: "request", value: function() {
            var t3 = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {};
            return o(t3, { xd: this.xd, xs: this.xs }, this.opts), new E(this.uri(), t3);
          } }, { key: "doWrite", value: function(t3, e3) {
            var n3 = this, r2 = this.request({ method: "POST", data: t3 });
            r2.on("success", e3), r2.on("error", function(t4) {
              n3.onError("xhr post error", t4);
            });
          } }, { key: "doPoll", value: function() {
            var t3 = this, e3 = this.request();
            e3.on("data", this.onData.bind(this)), e3.on("error", function(e4) {
              t3.onError("xhr poll error", e4);
            }), this.pollXhr = e3;
          } }]), n2;
        }(d), E = function(t2) {
          a(n2, t2);
          var e2 = f(n2);
          function n2(t3, r2) {
            var o2;
            return i(this, n2), o2 = e2.call(this), g(p(o2), r2), o2.opts = r2, o2.method = r2.method || "GET", o2.uri = t3, o2.async = r2.async !== false, o2.data = r2.data !== void 0 ? r2.data : null, o2.create(), o2;
          }
          return c(n2, [{ key: "create", value: function() {
            var t3 = this, e3 = m(this.opts, "agent", "enablesXDR", "pfx", "key", "passphrase", "cert", "ca", "ciphers", "rejectUnauthorized", "autoUnref");
            e3.xdomain = !!this.opts.xd, e3.xscheme = !!this.opts.xs;
            var r2 = this.xhr = new y(e3);
            try {
              r2.open(this.method, this.uri, this.async);
              try {
                if (this.opts.extraHeaders)
                  for (var o2 in r2.setDisableHeaderCheck && r2.setDisableHeaderCheck(true), this.opts.extraHeaders)
                    this.opts.extraHeaders.hasOwnProperty(o2) && r2.setRequestHeader(o2, this.opts.extraHeaders[o2]);
              } catch (t4) {
              }
              if (this.method === "POST")
                try {
                  r2.setRequestHeader("Content-type", "text/plain;charset=UTF-8");
                } catch (t4) {
                }
              try {
                r2.setRequestHeader("Accept", "*/*");
              } catch (t4) {
              }
              "withCredentials" in r2 && (r2.withCredentials = this.opts.withCredentials), this.opts.requestTimeout && (r2.timeout = this.opts.requestTimeout), this.hasXDR() ? (r2.onload = function() {
                t3.onLoad();
              }, r2.onerror = function() {
                t3.onError(r2.responseText);
              }) : r2.onreadystatechange = function() {
                r2.readyState === 4 && (r2.status === 200 || r2.status === 1223 ? t3.onLoad() : t3.setTimeoutFn(function() {
                  t3.onError(typeof r2.status == "number" ? r2.status : 0);
                }, 0));
              }, r2.send(this.data);
            } catch (e4) {
              return void this.setTimeoutFn(function() {
                t3.onError(e4);
              }, 0);
            }
            typeof document != "undefined" && (this.index = n2.requestsCount++, n2.requests[this.index] = this);
          } }, { key: "onSuccess", value: function() {
            this.emit("success"), this.cleanup();
          } }, { key: "onData", value: function(t3) {
            this.emit("data", t3), this.onSuccess();
          } }, { key: "onError", value: function(t3) {
            this.emit("error", t3), this.cleanup(true);
          } }, { key: "cleanup", value: function(t3) {
            if (this.xhr !== void 0 && this.xhr !== null) {
              if (this.hasXDR() ? this.xhr.onload = this.xhr.onerror = w : this.xhr.onreadystatechange = w, t3)
                try {
                  this.xhr.abort();
                } catch (t4) {
                }
              typeof document != "undefined" && delete n2.requests[this.index], this.xhr = null;
            }
          } }, { key: "onLoad", value: function() {
            var t3 = this.xhr.responseText;
            t3 !== null && this.onData(t3);
          } }, { key: "hasXDR", value: function() {
            return typeof XDomainRequest != "undefined" && !this.xs && this.enablesXDR;
          } }, { key: "abort", value: function() {
            this.cleanup();
          } }]), n2;
        }(v);
        if (E.requestsCount = 0, E.requests = {}, typeof document != "undefined") {
          if (typeof attachEvent == "function")
            attachEvent("onunload", S);
          else if (typeof addEventListener == "function") {
            addEventListener("onpagehide" in k ? "pagehide" : "unload", S, false);
          }
        }
        function S() {
          for (var t2 in E.requests)
            E.requests.hasOwnProperty(t2) && E.requests[t2].abort();
        }
        t.exports = _, t.exports.Request = E;
      }, function(t, e, n) {
        var r = n(12).PACKET_TYPES, o = typeof Blob == "function" || typeof Blob != "undefined" && Object.prototype.toString.call(Blob) === "[object BlobConstructor]", i = typeof ArrayBuffer == "function", s = function(t2, e2) {
          var n2 = new FileReader();
          return n2.onload = function() {
            var t3 = n2.result.split(",")[1];
            e2("b" + t3);
          }, n2.readAsDataURL(t2);
        };
        t.exports = function(t2, e2, n2) {
          var c, a = t2.type, u = t2.data;
          return o && u instanceof Blob ? e2 ? n2(u) : s(u, n2) : i && (u instanceof ArrayBuffer || (c = u, typeof ArrayBuffer.isView == "function" ? ArrayBuffer.isView(c) : c && c.buffer instanceof ArrayBuffer)) ? e2 ? n2(u instanceof ArrayBuffer ? u : u.buffer) : s(new Blob([u]), n2) : n2(r[a] + (u || ""));
        };
      }, function(t, e, n) {
        var r, o = n(12), i = o.PACKET_TYPES_REVERSE, s = o.ERROR_PACKET;
        typeof ArrayBuffer == "function" && (r = n(26));
        var c = function(t2, e2) {
          if (r) {
            var n2 = r.decode(t2);
            return a(n2, e2);
          }
          return { base64: true, data: t2 };
        }, a = function(t2, e2) {
          switch (e2) {
            case "blob":
              return t2 instanceof ArrayBuffer ? new Blob([t2]) : t2;
            case "arraybuffer":
            default:
              return t2;
          }
        };
        t.exports = function(t2, e2) {
          if (typeof t2 != "string")
            return { type: "message", data: a(t2, e2) };
          var n2 = t2.charAt(0);
          return n2 === "b" ? { type: "message", data: c(t2.substring(1), e2) } : i[n2] ? t2.length > 1 ? { type: i[n2], data: t2.substring(1) } : { type: i[n2] } : s;
        };
      }, function(t, e) {
        !function(t2) {
          "use strict";
          e.encode = function(e2) {
            var n, r = new Uint8Array(e2), o = r.length, i = "";
            for (n = 0; n < o; n += 3)
              i += t2[r[n] >> 2], i += t2[(3 & r[n]) << 4 | r[n + 1] >> 4], i += t2[(15 & r[n + 1]) << 2 | r[n + 2] >> 6], i += t2[63 & r[n + 2]];
            return o % 3 == 2 ? i = i.substring(0, i.length - 1) + "=" : o % 3 == 1 && (i = i.substring(0, i.length - 2) + "=="), i;
          }, e.decode = function(e2) {
            var n, r, o, i, s, c = 0.75 * e2.length, a = e2.length, u = 0;
            e2[e2.length - 1] === "=" && (c--, e2[e2.length - 2] === "=" && c--);
            var f = new ArrayBuffer(c), l = new Uint8Array(f);
            for (n = 0; n < a; n += 4)
              r = t2.indexOf(e2[n]), o = t2.indexOf(e2[n + 1]), i = t2.indexOf(e2[n + 2]), s = t2.indexOf(e2[n + 3]), l[u++] = r << 2 | o >> 4, l[u++] = (15 & o) << 4 | i >> 2, l[u++] = (3 & i) << 6 | 63 & s;
            return f;
          };
        }("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/");
      }, function(t, e, n) {
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function i(t2, e2, n2) {
          return (i = typeof Reflect != "undefined" && Reflect.get ? Reflect.get : function(t3, e3, n3) {
            var r2 = function(t4, e4) {
              for (; !Object.prototype.hasOwnProperty.call(t4, e4) && (t4 = f(t4)) !== null; )
                ;
              return t4;
            }(t3, e3);
            if (r2) {
              var o2 = Object.getOwnPropertyDescriptor(r2, e3);
              return o2.get ? o2.get.call(n3) : o2.value;
            }
          })(t2, e2, n2 || t2);
        }
        function s(t2, e2) {
          return (s = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function c(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = f(t2);
            if (e2) {
              var o2 = f(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return a(this, n2);
          };
        }
        function a(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return u(t2);
        }
        function u(t2) {
          if (t2 === void 0)
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
          return t2;
        }
        function f(t2) {
          return (f = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var l, p = n(11), h = n(0), y = /\n/g, d = /\\n/g, v = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && s(t3, e3);
          }(p2, t2);
          var e2, n2, r2, a2 = c(p2);
          function p2(t3) {
            var e3;
            return function(t4, e4) {
              if (!(t4 instanceof e4))
                throw new TypeError("Cannot call a class as a function");
            }(this, p2), (e3 = a2.call(this, t3)).query = e3.query || {}, l || (l = h.___eio = h.___eio || []), e3.index = l.length, l.push(e3.onData.bind(u(e3))), e3.query.j = e3.index, e3;
          }
          return e2 = p2, (n2 = [{ key: "supportsBinary", get: function() {
            return false;
          } }, { key: "doClose", value: function() {
            this.script && (this.script.onerror = function() {
            }, this.script.parentNode.removeChild(this.script), this.script = null), this.form && (this.form.parentNode.removeChild(this.form), this.form = null, this.iframe = null), i(f(p2.prototype), "doClose", this).call(this);
          } }, { key: "doPoll", value: function() {
            var t3 = this, e3 = document.createElement("script");
            this.script && (this.script.parentNode.removeChild(this.script), this.script = null), e3.async = true, e3.src = this.uri(), e3.onerror = function(e4) {
              t3.onError("jsonp poll error", e4);
            };
            var n3 = document.getElementsByTagName("script")[0];
            n3 ? n3.parentNode.insertBefore(e3, n3) : (document.head || document.body).appendChild(e3), this.script = e3, typeof navigator != "undefined" && /gecko/i.test(navigator.userAgent) && this.setTimeoutFn(function() {
              var t4 = document.createElement("iframe");
              document.body.appendChild(t4), document.body.removeChild(t4);
            }, 100);
          } }, { key: "doWrite", value: function(t3, e3) {
            var n3, r3 = this;
            if (!this.form) {
              var o2 = document.createElement("form"), i2 = document.createElement("textarea"), s2 = this.iframeId = "eio_iframe_" + this.index;
              o2.className = "socketio", o2.style.position = "absolute", o2.style.top = "-1000px", o2.style.left = "-1000px", o2.target = s2, o2.method = "POST", o2.setAttribute("accept-charset", "utf-8"), i2.name = "d", o2.appendChild(i2), document.body.appendChild(o2), this.form = o2, this.area = i2;
            }
            function c2() {
              a3(), e3();
            }
            this.form.action = this.uri();
            var a3 = function() {
              if (r3.iframe)
                try {
                  r3.form.removeChild(r3.iframe);
                } catch (t5) {
                  r3.onError("jsonp polling iframe removal error", t5);
                }
              try {
                var t4 = '<iframe src="javascript:0" name="' + r3.iframeId + '">';
                n3 = document.createElement(t4);
              } catch (t5) {
                (n3 = document.createElement("iframe")).name = r3.iframeId, n3.src = "javascript:0";
              }
              n3.id = r3.iframeId, r3.form.appendChild(n3), r3.iframe = n3;
            };
            a3(), t3 = t3.replace(d, "\\\n"), this.area.value = t3.replace(y, "\\n");
            try {
              this.form.submit();
            } catch (t4) {
            }
            this.iframe.attachEvent ? this.iframe.onreadystatechange = function() {
              r3.iframe.readyState === "complete" && c2();
            } : this.iframe.onload = c2;
          } }]) && o(e2.prototype, n2), r2 && o(e2, r2), p2;
        }(p);
        t.exports = v;
      }, function(t, e, n) {
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        function o(t2, e2) {
          for (var n2 = 0; n2 < e2.length; n2++) {
            var r2 = e2[n2];
            r2.enumerable = r2.enumerable || false, r2.configurable = true, "value" in r2 && (r2.writable = true), Object.defineProperty(t2, r2.key, r2);
          }
        }
        function i(t2, e2) {
          return (i = Object.setPrototypeOf || function(t3, e3) {
            return t3.__proto__ = e3, t3;
          })(t2, e2);
        }
        function s(t2) {
          var e2 = function() {
            if (typeof Reflect == "undefined" || !Reflect.construct)
              return false;
            if (Reflect.construct.sham)
              return false;
            if (typeof Proxy == "function")
              return true;
            try {
              return Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {
              })), true;
            } catch (t3) {
              return false;
            }
          }();
          return function() {
            var n2, r2 = a(t2);
            if (e2) {
              var o2 = a(this).constructor;
              n2 = Reflect.construct(r2, arguments, o2);
            } else
              n2 = r2.apply(this, arguments);
            return c(this, n2);
          };
        }
        function c(t2, e2) {
          if (e2 && (r(e2) === "object" || typeof e2 == "function"))
            return e2;
          if (e2 !== void 0)
            throw new TypeError("Derived constructors may only return object or undefined");
          return function(t3) {
            if (t3 === void 0)
              throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
            return t3;
          }(t2);
        }
        function a(t2) {
          return (a = Object.setPrototypeOf ? Object.getPrototypeOf : function(t3) {
            return t3.__proto__ || Object.getPrototypeOf(t3);
          })(t2);
        }
        var u = n(4), f = n(1), l = n(5), p = n(13), h = n(3).pick, y = n(29), d = y.WebSocket, v = y.usingBrowserWebSocket, b = y.defaultBinaryType, m = y.nextTick, g = typeof navigator != "undefined" && typeof navigator.product == "string" && navigator.product.toLowerCase() === "reactnative", k = function(t2) {
          !function(t3, e3) {
            if (typeof e3 != "function" && e3 !== null)
              throw new TypeError("Super expression must either be null or a function");
            t3.prototype = Object.create(e3 && e3.prototype, { constructor: { value: t3, writable: true, configurable: true } }), e3 && i(t3, e3);
          }(a2, t2);
          var e2, n2, r2, c2 = s(a2);
          function a2(t3) {
            var e3;
            return function(t4, e4) {
              if (!(t4 instanceof e4))
                throw new TypeError("Cannot call a class as a function");
            }(this, a2), (e3 = c2.call(this, t3)).supportsBinary = !t3.forceBase64, e3;
          }
          return e2 = a2, (n2 = [{ key: "name", get: function() {
            return "websocket";
          } }, { key: "doOpen", value: function() {
            if (this.check()) {
              var t3 = this.uri(), e3 = this.opts.protocols, n3 = g ? {} : h(this.opts, "agent", "perMessageDeflate", "pfx", "key", "passphrase", "cert", "ca", "ciphers", "rejectUnauthorized", "localAddress", "protocolVersion", "origin", "maxPayload", "family", "checkServerIdentity");
              this.opts.extraHeaders && (n3.headers = this.opts.extraHeaders);
              try {
                this.ws = v && !g ? e3 ? new d(t3, e3) : new d(t3) : new d(t3, e3, n3);
              } catch (t4) {
                return this.emit("error", t4);
              }
              this.ws.binaryType = this.socket.binaryType || b, this.addEventListeners();
            }
          } }, { key: "addEventListeners", value: function() {
            var t3 = this;
            this.ws.onopen = function() {
              t3.opts.autoUnref && t3.ws._socket.unref(), t3.onOpen();
            }, this.ws.onclose = this.onClose.bind(this), this.ws.onmessage = function(e3) {
              return t3.onData(e3.data);
            }, this.ws.onerror = function(e3) {
              return t3.onError("websocket error", e3);
            };
          } }, { key: "write", value: function(t3) {
            var e3 = this;
            this.writable = false;
            for (var n3 = function(n4) {
              var r4 = t3[n4], o2 = n4 === t3.length - 1;
              f.encodePacket(r4, e3.supportsBinary, function(t4) {
                var n5 = {};
                v || (r4.options && (n5.compress = r4.options.compress), e3.opts.perMessageDeflate && (typeof t4 == "string" ? Buffer.byteLength(t4) : t4.length) < e3.opts.perMessageDeflate.threshold && (n5.compress = false));
                try {
                  v ? e3.ws.send(t4) : e3.ws.send(t4, n5);
                } catch (t5) {
                }
                o2 && m(function() {
                  e3.writable = true, e3.emit("drain");
                }, e3.setTimeoutFn);
              });
            }, r3 = 0; r3 < t3.length; r3++)
              n3(r3);
          } }, { key: "onClose", value: function() {
            u.prototype.onClose.call(this);
          } }, { key: "doClose", value: function() {
            this.ws !== void 0 && (this.ws.close(), this.ws = null);
          } }, { key: "uri", value: function() {
            var t3 = this.query || {}, e3 = this.opts.secure ? "wss" : "ws", n3 = "";
            return this.opts.port && (e3 === "wss" && Number(this.opts.port) !== 443 || e3 === "ws" && Number(this.opts.port) !== 80) && (n3 = ":" + this.opts.port), this.opts.timestampRequests && (t3[this.opts.timestampParam] = p()), this.supportsBinary || (t3.b64 = 1), (t3 = l.encode(t3)).length && (t3 = "?" + t3), e3 + "://" + (this.opts.hostname.indexOf(":") !== -1 ? "[" + this.opts.hostname + "]" : this.opts.hostname) + n3 + this.opts.path + t3;
          } }, { key: "check", value: function() {
            return !(!d || "__initialize" in d && this.name === a2.prototype.name);
          } }]) && o(e2.prototype, n2), r2 && o(e2, r2), a2;
        }(u);
        t.exports = k;
      }, function(t, e, n) {
        var r = n(0), o = typeof Promise == "function" && typeof Promise.resolve == "function" ? function(t2) {
          return Promise.resolve().then(t2);
        } : function(t2, e2) {
          return e2(t2, 0);
        };
        t.exports = { WebSocket: r.WebSocket || r.MozWebSocket, usingBrowserWebSocket: true, defaultBinaryType: "arraybuffer", nextTick: o };
      }, function(t, e, n) {
        "use strict";
        function r(t2) {
          return (r = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t3) {
            return typeof t3;
          } : function(t3) {
            return t3 && typeof Symbol == "function" && t3.constructor === Symbol && t3 !== Symbol.prototype ? "symbol" : typeof t3;
          })(t2);
        }
        Object.defineProperty(e, "__esModule", { value: true }), e.reconstructPacket = e.deconstructPacket = void 0;
        var o = n(15);
        e.deconstructPacket = function(t2) {
          var e2 = [], n2 = t2.data, i = t2;
          return i.data = function t3(e3, n3) {
            if (!e3)
              return e3;
            if (o.isBinary(e3)) {
              var i2 = { _placeholder: true, num: n3.length };
              return n3.push(e3), i2;
            }
            if (Array.isArray(e3)) {
              for (var s = new Array(e3.length), c = 0; c < e3.length; c++)
                s[c] = t3(e3[c], n3);
              return s;
            }
            if (r(e3) === "object" && !(e3 instanceof Date)) {
              var a = {};
              for (var u in e3)
                e3.hasOwnProperty(u) && (a[u] = t3(e3[u], n3));
              return a;
            }
            return e3;
          }(n2, e2), i.attachments = e2.length, { packet: i, buffers: e2 };
        }, e.reconstructPacket = function(t2, e2) {
          return t2.data = function t3(e3, n2) {
            if (!e3)
              return e3;
            if (e3 && e3._placeholder)
              return n2[e3.num];
            if (Array.isArray(e3))
              for (var o2 = 0; o2 < e3.length; o2++)
                e3[o2] = t3(e3[o2], n2);
            else if (r(e3) === "object")
              for (var i in e3)
                e3.hasOwnProperty(i) && (e3[i] = t3(e3[i], n2));
            return e3;
          }(t2.data, e2), t2.attachments = void 0, t2;
        };
      }, function(t, e) {
        function n(t2) {
          t2 = t2 || {}, this.ms = t2.min || 100, this.max = t2.max || 1e4, this.factor = t2.factor || 2, this.jitter = t2.jitter > 0 && t2.jitter <= 1 ? t2.jitter : 0, this.attempts = 0;
        }
        t.exports = n, n.prototype.duration = function() {
          var t2 = this.ms * Math.pow(this.factor, this.attempts++);
          if (this.jitter) {
            var e2 = Math.random(), n2 = Math.floor(e2 * this.jitter * t2);
            t2 = (1 & Math.floor(10 * e2)) == 0 ? t2 - n2 : t2 + n2;
          }
          return 0 | Math.min(t2, this.max);
        }, n.prototype.reset = function() {
          this.attempts = 0;
        }, n.prototype.setMin = function(t2) {
          this.ms = t2;
        }, n.prototype.setMax = function(t2) {
          this.max = t2;
        }, n.prototype.setJitter = function(t2) {
          this.jitter = t2;
        };
      }]);
    });
  }
});

// src/client/index.ts
var import_socket = __toModule(require_socket_io_min());
window.startup = (options) => {
  var _a;
  const { routeName } = options;
  const iframe = document.getElementById("main");
  const win = iframe.contentWindow;
  if (!win) {
    return;
  }
  function log(...args) {
    console.log(`[${routeName}][socket ${socket.id}]`, ...args);
  }
  const socket = (0, import_socket.io)(options.url, { autoConnect: true });
  let storedState = options.state;
  win.acquireVsCodeApi = () => {
    return {
      getState() {
        return storedState;
      },
      setState(state) {
        log("setState", state);
        storedState = state;
        socket.emit("setState", state);
      },
      postMessage(message) {
        log("postMessage", message);
        socket.emit("postMessage", message);
      }
    };
  };
  socket.on("connect", () => {
    log(`register`);
    socket.emit("register", routeName);
  });
  socket.on("postMessage", (message) => {
    log("received postMessage", message);
    win.postMessage(message, "*");
  });
  let title = document.title;
  socket.on("title", (content) => {
    log("received title", content);
    title = content;
    const titleDom = document.querySelector("#title-content");
    if (titleDom)
      titleDom.textContent = content;
    document.title = content;
  });
  const colorMediaQuery = window.matchMedia("(prefers-color-scheme: dark)");
  let systemColorMode = colorMediaQuery.matches ? "dark" : "light";
  colorMediaQuery.addEventListener("change", (e) => {
    systemColorMode = e.matches ? "dark" : "light";
    setIconPath(systemColorMode);
  });
  let iconPaths;
  const setIconPath = (mode) => {
    if (!iconPaths)
      return;
    const titleIconDom = document.querySelector("#title-img");
    if (titleIconDom)
      titleIconDom.src = iconPaths[mode];
  };
  socket.on("iconPath", (paths) => {
    log("received iconPath", paths);
    iconPaths = paths;
    setIconPath(options.lightOrDarkMode === "system" ? systemColorMode : options.lightOrDarkMode);
  });
  const setHtml = (content) => {
    win.document.open();
    win.document.write(content);
    win.document.close();
    win.focus();
  };
  socket.on("html", (content) => {
    log("received html", content);
    setHtml(content);
  });
  const revealCover = document.getElementById("reveal-cover");
  revealCover.addEventListener("transitionend", () => {
    console.log("transitionend");
    revealCover.style.background = "";
  });
  const titleFlasher = {
    timer: void 0,
    start() {
      let reveal = true;
      titleFlasher.timer = setInterval(() => {
        document.title = reveal ? `[REVEAL] ${title}` : title;
        reveal = !reveal;
      }, 500);
    },
    stop() {
      if (titleFlasher.timer) {
        clearInterval(titleFlasher.timer);
        titleFlasher.timer = void 0;
        document.title = title;
      }
    }
  };
  socket.on("reveal", () => {
    if (!visible) {
      titleFlasher.start();
    }
    const flashTimes = 2;
    const intervalTime = 0.6;
    revealCover.style.transition = `background ${intervalTime / 2}s`;
    revealCover.style.display = "block";
    for (let i = 0; i < flashTimes; i++) {
      setTimeout(() => {
        revealCover.style.background = "rgba(0,0,0, 0.7)";
      }, i * intervalTime * 1e3);
    }
    setTimeout(() => {
      revealCover.style.display = "none";
      revealCover.style.background = "";
    }, flashTimes * intervalTime * 1e3);
  });
  socket.on("disconnect", () => {
    titleFlasher.stop();
  });
  const close = () => {
    log(`close`);
    setHtml("CLOSED");
    win.close();
    if (!options.debug) {
      window.close();
    }
  };
  socket.on("dispose", () => {
    titleFlasher.stop();
    close();
  });
  win.focus();
  let visible = true;
  document.addEventListener("visibilitychange", () => {
    visible = !document.hidden;
    if (visible) {
      titleFlasher.stop();
    }
    socket.emit("visible", visible);
  });
  (_a = document.querySelector("#title .close")) == null ? void 0 : _a.addEventListener("click", () => {
    socket.emit("dispose");
    close();
  });
};
/*!
 * Socket.IO v4.2.0
 * (c) 2014-2021 Guillermo Rauch
 * Released under the MIT License.
 */
